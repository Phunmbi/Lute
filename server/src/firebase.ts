import admin, {firestore, ServiceAccount} from 'firebase-admin';
import serviceAccount from "../../serviceAccountKey.json";


admin.initializeApp({
	credential: admin.credential.cert(serviceAccount as ServiceAccount),
	databaseURL: 'https://construyo-coding-challenge.firebaseio.com'
})

export const db: firestore.Firestore = firestore();

