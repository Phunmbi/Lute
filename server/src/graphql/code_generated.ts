import { GraphQLResolveInfo, GraphQLScalarType, GraphQLScalarTypeConfig } from 'graphql';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type RequireFields<T, K extends keyof T> = Omit<T, K> & { [P in K]-?: NonNullable<T[P]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Date: any;
};

export type Address = {
  __typename?: 'Address';
  city?: Maybe<Scalars['String']>;
  country?: Maybe<Scalars['String']>;
  street?: Maybe<Scalars['String']>;
  zip?: Maybe<Scalars['String']>;
};

export type AddressInput = {
  city?: InputMaybe<Scalars['String']>;
  country?: InputMaybe<Scalars['String']>;
  street?: InputMaybe<Scalars['String']>;
  zip?: InputMaybe<Scalars['String']>;
};

export type AllOrdersResponse = {
  __typename?: 'AllOrdersResponse';
  count?: Maybe<Scalars['Int']>;
  cursor?: Maybe<Scalars['String']>;
  limit?: Maybe<Scalars['Int']>;
  orders?: Maybe<Array<OrderResponse>>;
};

export type Edge = {
  __typename?: 'Edge';
  cursor?: Maybe<Scalars['String']>;
  node?: Maybe<OrderResponse>;
};

export type Mutation = {
  __typename?: 'Mutation';
  createOrder: OrderResponse;
  updateOrder: OrderResponse;
};


export type MutationCreateOrderArgs = {
  orderRequest: OrderRequestBody;
};


export type MutationUpdateOrderArgs = {
  id: Scalars['String'];
  orderRequest: OrderRequestBody;
};

export type OrderRequestBody = {
  address?: InputMaybe<AddressInput>;
  bookingDate?: InputMaybe<Scalars['Date']>;
  customer?: InputMaybe<UserInput>;
  title?: InputMaybe<Scalars['String']>;
};

export type OrderResponse = {
  __typename?: 'OrderResponse';
  address?: Maybe<Address>;
  bookingDate?: Maybe<Scalars['Date']>;
  customer?: Maybe<User>;
  title?: Maybe<Scalars['String']>;
  uid: Scalars['String'];
};

export type OrdersConnection = {
  __typename?: 'OrdersConnection';
  edges?: Maybe<Array<Maybe<Edge>>>;
  pageInfo?: Maybe<PageInfo>;
  totalCount?: Maybe<Scalars['Int']>;
};

export type PageInfo = {
  __typename?: 'PageInfo';
  endCursor: Scalars['String'];
  hasNextPage: Scalars['Boolean'];
  hasPreviousPage: Scalars['Boolean'];
  startCursor: Scalars['String'];
};

export type Query = {
  __typename?: 'Query';
  allOrders?: Maybe<OrdersConnection>;
  order?: Maybe<OrderResponse>;
};


export type QueryAllOrdersArgs = {
  after?: InputMaybe<Scalars['String']>;
  before?: InputMaybe<Scalars['String']>;
  first?: InputMaybe<Scalars['Int']>;
  last?: InputMaybe<Scalars['Int']>;
};


export type QueryOrderArgs = {
  id: Scalars['String'];
};

export type User = {
  __typename?: 'User';
  email?: Maybe<Scalars['String']>;
  name?: Maybe<Scalars['String']>;
  phone?: Maybe<Scalars['String']>;
};

export type UserInput = {
  email?: InputMaybe<Scalars['String']>;
  name?: InputMaybe<Scalars['String']>;
  phone?: InputMaybe<Scalars['String']>;
};



export type ResolverTypeWrapper<T> = Promise<T> | T;


export type ResolverWithResolve<TResult, TParent, TContext, TArgs> = {
  resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> = ResolverFn<TResult, TParent, TContext, TArgs> | ResolverWithResolve<TResult, TParent, TContext, TArgs>;

export type ResolverFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => Promise<TResult> | TResult;

export type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => AsyncIterable<TResult> | Promise<AsyncIterable<TResult>>;

export type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

export interface SubscriptionSubscriberObject<TResult, TKey extends string, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<{ [key in TKey]: TResult }, TParent, TContext, TArgs>;
  resolve?: SubscriptionResolveFn<TResult, { [key in TKey]: TResult }, TContext, TArgs>;
}

export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>;
  resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>;
}

export type SubscriptionObject<TResult, TKey extends string, TParent, TContext, TArgs> =
  | SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs>
  | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>;

export type SubscriptionResolver<TResult, TKey extends string, TParent = {}, TContext = {}, TArgs = {}> =
  | ((...args: any[]) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>)
  | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>;

export type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (
  parent: TParent,
  context: TContext,
  info: GraphQLResolveInfo
) => Maybe<TTypes> | Promise<Maybe<TTypes>>;

export type IsTypeOfResolverFn<T = {}, TContext = {}> = (obj: T, context: TContext, info: GraphQLResolveInfo) => boolean | Promise<boolean>;

export type NextResolverFn<T> = () => Promise<T>;

export type DirectiveResolverFn<TResult = {}, TParent = {}, TContext = {}, TArgs = {}> = (
  next: NextResolverFn<TResult>,
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

/** Mapping between all available schema types and the resolvers types */
export type ResolversTypes = {
  Address: ResolverTypeWrapper<Address>;
  AddressInput: AddressInput;
  AllOrdersResponse: ResolverTypeWrapper<AllOrdersResponse>;
  Boolean: ResolverTypeWrapper<Scalars['Boolean']>;
  Date: ResolverTypeWrapper<Scalars['Date']>;
  Edge: ResolverTypeWrapper<Edge>;
  Int: ResolverTypeWrapper<Scalars['Int']>;
  Mutation: ResolverTypeWrapper<{}>;
  OrderRequestBody: OrderRequestBody;
  OrderResponse: ResolverTypeWrapper<OrderResponse>;
  OrdersConnection: ResolverTypeWrapper<OrdersConnection>;
  PageInfo: ResolverTypeWrapper<PageInfo>;
  Query: ResolverTypeWrapper<{}>;
  String: ResolverTypeWrapper<Scalars['String']>;
  User: ResolverTypeWrapper<User>;
  UserInput: UserInput;
};

/** Mapping between all available schema types and the resolvers parents */
export type ResolversParentTypes = {
  Address: Address;
  AddressInput: AddressInput;
  AllOrdersResponse: AllOrdersResponse;
  Boolean: Scalars['Boolean'];
  Date: Scalars['Date'];
  Edge: Edge;
  Int: Scalars['Int'];
  Mutation: {};
  OrderRequestBody: OrderRequestBody;
  OrderResponse: OrderResponse;
  OrdersConnection: OrdersConnection;
  PageInfo: PageInfo;
  Query: {};
  String: Scalars['String'];
  User: User;
  UserInput: UserInput;
};

export type AddressResolvers<ContextType = any, ParentType extends ResolversParentTypes['Address'] = ResolversParentTypes['Address']> = {
  city?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  country?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  street?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  zip?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type AllOrdersResponseResolvers<ContextType = any, ParentType extends ResolversParentTypes['AllOrdersResponse'] = ResolversParentTypes['AllOrdersResponse']> = {
  count?: Resolver<Maybe<ResolversTypes['Int']>, ParentType, ContextType>;
  cursor?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  limit?: Resolver<Maybe<ResolversTypes['Int']>, ParentType, ContextType>;
  orders?: Resolver<Maybe<Array<ResolversTypes['OrderResponse']>>, ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export interface DateScalarConfig extends GraphQLScalarTypeConfig<ResolversTypes['Date'], any> {
  name: 'Date';
}

export type EdgeResolvers<ContextType = any, ParentType extends ResolversParentTypes['Edge'] = ResolversParentTypes['Edge']> = {
  cursor?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  node?: Resolver<Maybe<ResolversTypes['OrderResponse']>, ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type MutationResolvers<ContextType = any, ParentType extends ResolversParentTypes['Mutation'] = ResolversParentTypes['Mutation']> = {
  createOrder?: Resolver<ResolversTypes['OrderResponse'], ParentType, ContextType, RequireFields<MutationCreateOrderArgs, 'orderRequest'>>;
  updateOrder?: Resolver<ResolversTypes['OrderResponse'], ParentType, ContextType, RequireFields<MutationUpdateOrderArgs, 'id' | 'orderRequest'>>;
};

export type OrderResponseResolvers<ContextType = any, ParentType extends ResolversParentTypes['OrderResponse'] = ResolversParentTypes['OrderResponse']> = {
  address?: Resolver<Maybe<ResolversTypes['Address']>, ParentType, ContextType>;
  bookingDate?: Resolver<Maybe<ResolversTypes['Date']>, ParentType, ContextType>;
  customer?: Resolver<Maybe<ResolversTypes['User']>, ParentType, ContextType>;
  title?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  uid?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type OrdersConnectionResolvers<ContextType = any, ParentType extends ResolversParentTypes['OrdersConnection'] = ResolversParentTypes['OrdersConnection']> = {
  edges?: Resolver<Maybe<Array<Maybe<ResolversTypes['Edge']>>>, ParentType, ContextType>;
  pageInfo?: Resolver<Maybe<ResolversTypes['PageInfo']>, ParentType, ContextType>;
  totalCount?: Resolver<Maybe<ResolversTypes['Int']>, ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type PageInfoResolvers<ContextType = any, ParentType extends ResolversParentTypes['PageInfo'] = ResolversParentTypes['PageInfo']> = {
  endCursor?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  hasNextPage?: Resolver<ResolversTypes['Boolean'], ParentType, ContextType>;
  hasPreviousPage?: Resolver<ResolversTypes['Boolean'], ParentType, ContextType>;
  startCursor?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type QueryResolvers<ContextType = any, ParentType extends ResolversParentTypes['Query'] = ResolversParentTypes['Query']> = {
  allOrders?: Resolver<Maybe<ResolversTypes['OrdersConnection']>, ParentType, ContextType, Partial<QueryAllOrdersArgs>>;
  order?: Resolver<Maybe<ResolversTypes['OrderResponse']>, ParentType, ContextType, RequireFields<QueryOrderArgs, 'id'>>;
};

export type UserResolvers<ContextType = any, ParentType extends ResolversParentTypes['User'] = ResolversParentTypes['User']> = {
  email?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  name?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  phone?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
};

export type Resolvers<ContextType = any> = {
  Address?: AddressResolvers<ContextType>;
  AllOrdersResponse?: AllOrdersResponseResolvers<ContextType>;
  Date?: GraphQLScalarType;
  Edge?: EdgeResolvers<ContextType>;
  Mutation?: MutationResolvers<ContextType>;
  OrderResponse?: OrderResponseResolvers<ContextType>;
  OrdersConnection?: OrdersConnectionResolvers<ContextType>;
  PageInfo?: PageInfoResolvers<ContextType>;
  Query?: QueryResolvers<ContextType>;
  User?: UserResolvers<ContextType>;
};

