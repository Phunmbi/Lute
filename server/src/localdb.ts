export const orders = [{
	"address": {
		"city": "Berlin",
		"country": "Germany",
		"street": "Wriezener Str. 12",
		"zip": "13055"
	},
	"bookingDate": 1554284950000,
	"customer": {
		"email": "emad.alam@construyo.de",
		"name": "Emad Alam",
		"phone": "015252098067"
	},
	"title": "Test Order 1",
	"uid": "hKlIKPoZc2xCKGTUKZK2"
}]

export const user = {
	"email": "coding-challenge@construyo",
	"name": "Construyo Coding Challenge",
	"phone": "0123456789",
	"uid": "5iEm1HvIxubLaiKO4yj0Npmvq0F2"
}