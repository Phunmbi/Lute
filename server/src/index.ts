import express, {Express} from 'express';
import {createServer} from 'http';
import createApolloServer from './apollo-server';
import {db} from './firebase';

const app: Express = express();

async function start() {
	const httpServer = createServer(app);
	const apolloServer = await createApolloServer(db, httpServer, app);
	
	app.listen('4000', () => {
		console.log('GraphQL API Server listening on 4000');
		console.log(`sandbox: http://localhost:4000${apolloServer.graphqlPath}`);
	});
}

start().catch((e) => {
	console.error(e);
});
