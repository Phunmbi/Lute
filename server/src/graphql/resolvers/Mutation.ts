import {MutationCreateOrderArgs, MutationResolvers} from '../code_generated';
import {ResolverContext} from '../../resolvers';
import OrdersService from "../../services/orders/orders";

const mutationOrderResolver: MutationResolvers<ResolverContext> = {
	createOrder: async (_, args: MutationCreateOrderArgs, context: ResolverContext) => {
		return OrdersService(context.db).createOrder(args)
	},
	
	updateOrder: async (_, args, context: ResolverContext) => {
		return OrdersService(context.db).updateOrder(args)
	},
};

export default mutationOrderResolver;
