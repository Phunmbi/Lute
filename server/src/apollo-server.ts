import {ApolloServerPluginDrainHttpServer} from "apollo-server-core"
import {ApolloServer, ExpressContext} from "apollo-server-express"
import { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader"
import { loadSchemaSync } from "@graphql-tools/load"
import { addResolversToSchema } from "@graphql-tools/schema"
import  express from "express"
import { Server } from "http"
import resolvers from "./resolvers";

const GRAPHQL_SCHEMA_PATH = '../schema.graphql'
const SCHEMA = loadSchemaSync(GRAPHQL_SCHEMA_PATH, {
	loaders: [new GraphQLFileLoader()],
})

export async function createApolloServer(
	db: String,
	httpServer: Server,
	app: express.Application
): Promise<ApolloServer<ExpressContext>> {
	const server = new ApolloServer({
		schema: addResolversToSchema({
			schema: SCHEMA, resolvers
		}),
		context: () => ({ db }),
		plugins: [
			ApolloServerPluginDrainHttpServer({ httpServer }),
		],
	})
	
	await server.start()
	server.applyMiddleware({ app })
	return server
}