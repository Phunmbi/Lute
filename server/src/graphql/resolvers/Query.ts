import { QueryResolvers } from '../code_generated';
import OrdersService from '../../services/orders';
import { ResolverContext } from '../../resolvers';

const queryOrderResolver: QueryResolvers<ResolverContext> = {
  getOrder: async (_, { id }, { db }) => OrdersService(db).getOrder(id),
};

export default queryOrderResolver;
