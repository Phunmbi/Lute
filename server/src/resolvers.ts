import {firestore} from 'firebase-admin';
import Query from './graphql/resolvers/Query';
import {Resolvers} from './graphql/code_generated';
import Mutation from "./graphql/resolvers/Mutation";
import Firestore = firestore.Firestore;

export interface ResolverContext {
	db: Firestore
	user: { uid: string }
}

const resolvers: Resolvers<ResolverContext> = {
	Query,
	Mutation
};

export default resolvers;
