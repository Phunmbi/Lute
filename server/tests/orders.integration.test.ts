import { ApolloServer } from "apollo-server-express";
import resolvers from "../src/resolvers";
import { SCHEMA } from "../src/apollo-server";
import GET_ALL_ORDERS from "client/src/graphql/allOrdersQuery";
import GET_SINGLE_ORDER from "client/src/graphql/singleOrderQuery";
import { CREATE_ORDER } from "client/src/graphql/createOrderMutation";
import { inputOrder } from "../src/localdb";

const server = new ApolloServer({
	schema: SCHEMA,
	resolvers,
	context: () => ({ user: { id: 1, email: "a@a.a" } }),
});

describe("Queries", () => {
	it("should fetch all orders", async () => {
		const res = await server.executeOperation({
			query: GET_ALL_ORDERS,
			variables: { first: 10 },
		});

		expect(res.errors).toBeUndefined();
		expect(res).toMatchSnapshot();
	});

	it("should fetch a single Order", async () => {
		const res = await server.executeOperation({
			query: GET_SINGLE_ORDER,
			variables: { id: "CeAmCcbx0JJ5QNWgFTDW" },
		});

		expect(res.errors).toBeUndefined();
		expect(res).toMatchSnapshot();
	});
});

describe("Mutations", () => {
	it("should create order", async () => {
		const res = await server.executeOperation({
			query: CREATE_ORDER,
			variables: { orderRequest: { ...inputOrder[0] } },
		});

		// expect(res.errors).toBeUndefined();
		expect(res).toMatchSnapshot();
	});
});
