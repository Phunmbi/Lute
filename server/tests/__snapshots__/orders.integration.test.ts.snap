// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutations should create order 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Cannot return null for non-nullable field Mutation.createOrder.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches all orders 1`] = `
Object {
  "data": Object {
    "allOrders": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries should fetch a single Order 1`] = `
Object {
  "data": Object {
    "order": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries should fetch all orders 1`] = `
Object {
  "data": Object {
    "allOrders": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
